lumicks/pylake/kymotracker/kymotracker.py

```diff
@@ -0,0 +1,174 @@
+    lines = points_to_line_segments(
+        data,
+        peaks,
+        kymo_score(vel=vel, sigma=sigma, diffusion=diffusion),
+        window=window,
+        sigma_cutoff=sigma_cutoff,
+    )
+
+    return [line.with_offset(rect[0][0], rect[0][1]) for line in lines] if rect else lines
+
+
+def track_lines(data, line_width, max_lines, start_threshold=0.005, continuation_threshold=0.005, angle_weight=10.0,
+                rect=None):
+    """Track particles on an image using an algorithm that looks for line-like structures.
+
+    Note: This is ALPHA functionality. It has not been tested in a sufficient number of cases yet, and the API is still
+    subject to change without a prior deprecation notice.
+
+    This function tracks particles in an image. It takes a pixel image, and traces lines on it. These lines can
+    subsequently be refined and/or used to extract intensities or other parameters.
+
+    This method is based on sections 1, 2 and 3 from [1]. This method attempts to find lines purely based on
+    differential geometric considerations. It blurs the image based with a user specified line width and then attempts
+    to find curvilinear sections. Based on eigenvalue decomposition of the local Hessian it finds the principal
+    direction of the line. It then computes subpixel accurate positions by computing the maximum perpendicular
+    to the line using local Taylor expansions. Pixels where this subpixel position estimate falls within the pixel are
+    considered lines. The initial step leads to lines with a width of one. These lines are then traced. When ambiguity
+    arises, on which point to connect next, a score comprised of the distance to the next subpixel minimum and angle
+    between the successive normal vectors is computed. The candidate with the lowest score is then selected.
+
+    For more information, please refer to the paper.
+
+    Parameters
+    ----------
+    data : array_like
+        N by M image containing a single color channel.
+    line_width : float
+        Expected line width in pixels.
+    max_lines : int
+        Maximum number of lines to trace.
+    start_threshold : float
+        Threshold for the value of the derivative.
+    continuation_threshold : float
+        Derivative threshold for the continuation of a line.
+    angle_weight: float
+        Factor which determines how the angle between normals needs to be weighted relative to distance.
+        High values push for straighter lines. Weighting occurs according to distance + angle_weight * angle difference
+    rect : tuple of two pixel coordinates
+        Only perform tracking over a subset of the image. Pixel coordinates should be given as:
+        ((min_time, min_coord), (max_time, max_coord)).
+
+    References
+    ----------
+        [1] Steger, C. (1998). An unbiased detector of curvilinear structures. IEEE Transactions on pattern analysis and
+        machine intelligence, 20(2), 113-125.
```

Comment:
Four spaces to the left like the one above.

Reply:
👍